### PR TITLE
refactor(ui): update global page context

### PR DESCRIPTION
**Pull request by simeng-li**
## Summary

Refactor the page context. Provide a new `usePageContext` hook that handles all page context setup logic.  Which returns the context provider as well as the global context.

Also, create the HOF `renderWithContext` for testing convenience to provide the pageContext on top of the React Component to be tested. 

Add global terms of use Agreement state to the global context for future use. As terms of use will be controlled cross-page components. 

## Linear Issue Reference

log-2160

## Testing

test locally update the ut
@logto-io/eng

### CODE REVIEW

File: packages/ui/src/__mocks__/RenderWithContext/index.tsx [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import { render, Queries, queries, RenderOptions } from '@testing-library/react';
+import React, { ReactElement } from 'react';
+
+import ContextProvider from './ContextProvider';
+
+const renderWithContext = <
```
Comment (xiaoyijun) [inline packages/ui/src/__mocks__/RenderWithContext/index.tsx:6]: How about `renderWithPageContext`?

File: packages/ui/src/components/AppContent/index.module.scss [hunk 2]
```diff
@@ -42,7 +42,7 @@ $font-family-small: 'PingFang SC', 'SF Pro Text', 'Siyuan Heiti', 'Roboto';
   --color-layer: linear-gradient(0deg, rgba(202, 190, 255, 14%), rgba(202, 190, 255, 14%)), linear-gradient(0deg, rgba(196, 199, 199, 2%), rgba(196, 199, 199, 2%)), #191c1d;
   --color-error: #dd3730;
   --color-toast: rgba(247, 248, 248, 80%);
-  --color-overlay: rgba(247, 248, 248, 80%);
+  --color-overlay: rgba(247, 248, 248, 16%);
```
Comment (xiaoyijun) [inline packages/ui/src/components/AppContent/index.module.scss:45]: nit: Why this color will change here, maybe this belongs to another PR.

File: packages/ui/src/containers/PasscodeValidation/index.tsx [hunk 3]
```diff
@@ -102,7 +102,7 @@ const PasscodeValidation = ({ type, method, className, target }: Props) => {
     <form className={classNames(styles.form, className)}>
       <Passcode
         name="passcode"
-        className={classNames(styles.field, error && styles.withError)}
+        className={classNames(styles.inputField, error && styles.withError)}
```
Comment (xiaoyijun) [inline packages/ui/src/containers/PasscodeValidation/index.tsx:105]: nit: ditto. belongs to another PR?

**Review by xiaoyijun — APPROVED** [review]
Overall LGTM.

File: packages/ui/src/hooks/use-page-context.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+      setTermsAgreement,
+      setShowTermsModal,
+      setExperienceSettings,
+    }),
+    [experienceSettings, loading, showTermsModal, termsAgreement, toast]
+  );
+
+  return {
+    context,
+    Provider: PageContext.Provider,
```
Comment (wangsijie) [inline packages/ui/src/hooks/use-page-context.ts:60]: nit: looks uncessary to me, since `PageContext` is already exported.

**Review by wangsijie — APPROVED** [review]

File: packages/ui/src/components/AppContent/index.module.scss (hunk 2 above)
Comment (simeng-li) [reply packages/ui/src/components/AppContent/index.module.scss:45]: yes, it's a bug found when testing this pr.  Hard to find a proper PR to update this simple issue, so leave it here.

File: packages/ui/src/hooks/use-page-context.ts [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  showTermsModal: boolean;
+  experienceSettings: SignInExperienceSettings | undefined;
+  setToast: (message: string) => void;
+  setLoading: (loading: boolean) => void;
+  setTermsAgreement: (termsAgreement: boolean) => void;
+  setShowTermsModal: (showTermsModal: boolean) => void;
+  setExperienceSettings: (settings: SignInExperienceSettings) => void;
+};
+
+const NOOP = () => {
```
Comment (demonzoo) [inline packages/ui/src/hooks/use-page-context.ts:18]: nit: our convention encourages only camelcase I remember? So just `noop`

File: packages/ui/src/containers/PasscodeValidation/index.tsx (hunk 3 above)
Comment (simeng-li) [reply packages/ui/src/containers/PasscodeValidation/index.tsx:105]: minor update as above, I'll split it next time.

File: packages/ui/src/hooks/use-page-context.ts (hunk 4 above)
Comment (simeng-li) [reply packages/ui/src/hooks/use-page-context.ts:60]: true, just thought it might follow the hook usage conventions when people want to get the Provider and context instance.

File: packages/ui/src/__mocks__/RenderWithContext/index.tsx (hunk 1 above)
Comment (simeng-li) [reply packages/ui/src/__mocks__/RenderWithContext/index.tsx]: updated

File: packages/ui/src/hooks/use-page-context.ts (hunk 5 above)
Comment (simeng-li) [reply packages/ui/src/hooks/use-page-context.ts]: updated

File: packages/ui/src/hooks/use-page-context.ts (hunk 4 above)
Comment (wangsijie) [reply packages/ui/src/hooks/use-page-context.ts:60]: OK
